core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -0,0 +1,736 @@
+            } catch (Exception ex) {
+                if (!isCancelled()) {
+                    logger.warn("Current task for topic-partition {} received error but it will be scheduled. " +
+                            "Reason: {}", topicIdPartition, ex.getMessage());
+                }
+            }
+        }
+
+        public String toString() {
+            return this.getClass().toString() + "[" + topicIdPartition + "]";
+        }
+    }
+
+    long findHighestRemoteOffset(TopicIdPartition topicIdPartition) throws RemoteStorageException {
+        Optional<Long> offset = Optional.empty();
+        Optional<UnifiedLog> maybeLog = fetchLog.apply(topicIdPartition.topicPartition());
+        if (maybeLog.isPresent()) {
+            UnifiedLog log = maybeLog.get();
+            Option<LeaderEpochFileCache> maybeLeaderEpochFileCache = log.leaderEpochCache();
+            if (maybeLeaderEpochFileCache.isDefined()) {
+                LeaderEpochFileCache cache = maybeLeaderEpochFileCache.get();
+                OptionalInt epoch = cache.latestEpoch();
+                while (!offset.isPresent() && epoch.isPresent()) {
+                    offset = remoteLogMetadataManager.highestOffsetForEpoch(topicIdPartition, epoch.getAsInt());
+                    epoch = cache.previousEpoch(epoch.getAsInt());
+                }
+            }
+        }
+
+        return offset.orElse(-1L);
+    }
+
+    void doHandleLeaderOrFollowerPartitions(TopicIdPartition topicPartition,
+                                            Consumer<RLMTask> convertToLeaderOrFollower) {
+        RLMTaskWithFuture rlmTaskWithFuture = leaderOrFollowerTasks.computeIfAbsent(topicPartition,
+                topicIdPartition -> {
+                    RLMTask task = new RLMTask(topicIdPartition);
+                    // set this upfront when it is getting initialized instead of doing it after scheduling.
+                    convertToLeaderOrFollower.accept(task);
+                    LOGGER.info("Created a new task: {} and getting scheduled", task);
+                    ScheduledFuture future = rlmScheduledThreadPool.scheduleWithFixedDelay(task, 0, delayInMs, TimeUnit.MILLISECONDS);
+                    return new RLMTaskWithFuture(task, future);
+                }
+        );
+        convertToLeaderOrFollower.accept(rlmTaskWithFuture.rlmTask);
+    }
+
+    static class RLMTaskWithFuture {
+
+        private final RLMTask rlmTask;
+        private final Future<?> future;
+
+        RLMTaskWithFuture(RLMTask rlmTask, Future<?> future) {
+            this.rlmTask = rlmTask;
+            this.future = future;
+        }
+
+        public void cancel() {
+            rlmTask.cancel();
+            try {
+                future.cancel(true);
+            } catch (Exception ex) {
+                LOGGER.error("Error occurred while canceling the task: {}", rlmTask, ex);
+            }
+        }
+
+    }
+
+    static class InMemoryLeaderEpochCheckpoint implements LeaderEpochCheckpoint {
```

Comment:
This is not necessary, and use `org.apache.kafka.storage.internals.checkpoint.InMemoryLeaderEpochCheckpoint` instead.